process/blocksPool.go

```diff
@@ -2,203 +2,114 @@ package process
-
-	indexesMap := make(map[uint32]uint64, len(lastCheckpoint.LastRounds))
-	for shardID, round := range lastCheckpoint.LastRounds {
-		indexesMap[shardID] = round
-	}
-	bp.previousIndexesMap = indexesMap
+// UpdateMetaState will triger meta state update from base data pool
+func (bp *blocksPool) UpdateMetaState(checkpoint *data.BlockCheckpoint) error {
+	return bp.dataPool.UpdateMetaState(checkpoint)
 }
 
-// Put will put value into storer
-func (bp *blocksPool) Put(key []byte, value []byte) error {
-	return bp.storer.Put(key, value)
+// GetLastCheckpoint returns last checkpoint data
+func (bp *blocksPool) GetLastCheckpoint() (*data.BlockCheckpoint, error) {
+	return bp.dataPool.GetLastCheckpoint()
 }
 
-// Get will get value from storer
+// Get will trigger data pool get operation
 func (bp *blocksPool) Get(key []byte) ([]byte, error) {
-	return bp.storer.Get(key)
-}
-
-// UpdateMetaState will update internal meta state
-func (bp *blocksPool) UpdateMetaState(checkpoint *data.BlockCheckpoint) error {
-	index, ok := checkpoint.LastRounds[core.MetachainShardId]
-	if !ok {
-		index = initIndex
-	}
-
-	if index >= bp.firstCommitableBlock {
-		err := bp.setCheckpoint(checkpoint)
-		if err != nil {
-			return fmt.Errorf("%w, failed to set checkpoint", err)
-		}
-	}
-
-	err := bp.pruneStorer(index)
+	data, err := bp.dataPool.Get(key)
 	if err != nil {
-		return fmt.Errorf("%w, failed to prune storer", err)
+		return nil, fmt.Errorf("failed to get from data pool: %w", err)
 	}
 
-	return nil
-}
-
-func (bp *blocksPool) pruneStorer(index uint64) error {
-	if index%bp.cleanupInterval != 0 {
-		return nil
-	}
-
-	return bp.storer.Prune(index)
+	return data, nil
 }
 
 // PutBlock will put the provided outport block data to the pool
-func (bp *blocksPool) PutBlock(hash []byte, value []byte, newIndex uint64, shardID uint32) error {
-	bp.mutMap.Lock()
-	defer bp.mutMap.Unlock()
-
-	previousIndex, ok := bp.previousIndexesMap[shardID]
-	if !ok {
-		bp.previousIndexesMap[shardID] = initIndex
-
-		err := bp.storer.Put(hash, value)
-		if err != nil {
-			return err
-		}
-
-		bp.previousIndexesMap[shardID] = newIndex
-
-		return nil
+func (bp *blocksPool) PutBlock(hash []byte, outportBlock OutportBlockHandler) error {
+	shardID := outportBlock.GetShardID()
+	currentIndex, err := outportBlock.GetHeaderNonce()
+	if err != nil {
+		return err
 	}
 
-	if !bp.shouldPutBlockData(previousIndex) {
-		return fmt.Errorf("%w: not within required delta, previous index %d, new index %d",
-			ErrFailedToPutBlockDataToPool, previousIndex, newIndex)
+	previousIndex := outportBlock.GetHighestFinalBlockNonce()
+	isHigherIndex := currentIndex >= previousIndex
+	if !isHigherIndex {
+		return fmt.Errorf("%w: new meta index should be higher than previous, previous index %d, new index %d",
+			ErrFailedToPutBlockDataToPool, previousIndex, currentIndex)
 	}
 
-	err := bp.storer.Put(hash, value)
+	outportBlockBytes, err := bp.marshaller.Marshal(outportBlock)
```

Comment:
Maybe we don't need to marshal to add in the pool
Would need to marshal before storing though.
Just to keep in mind for later.

Reply:
added todo for this